src/Orleans/Messaging/Message.cs

```diff
@@ -278,13 +280,29 @@ public ResponseTypes Result
             set {  Headers.Result = value; }
         }
 
-        public DateTime? Expiration
+        public TimeSpan? Expiration
         {
-            get { return Headers.Expiration; }
+            get
+            {
+                var now = DateTime.UtcNow;
+                var elapsed = now - localCreationTime;
+                Headers.Expiration = Headers.Expiration - elapsed;
+                localCreationTime = now;
```

Comment:
since you are doing this, maybe the field should be renamed?
How about `localTimeToLiveStartTime`? (since you are also renaming the Expiration property to TimeToLive).